### ISSUE TITLE
Auto terminate unused cloud(AWS,Google,Azure) resources 

### ISSUE THREAD

**Issue by phanikumv** [astronomer/astronomer-providers#44]
Create a utility script to auto terminate the cloud resources, if not used within the last 1 hr. This will improve our resource usage efficiency and reduce costs.

For instance, we can create a Lambda function on AWS which checks whether a particular resource(say, for example an EMR cluster) is unused for the last 1 hr and then terminate it automatically.

**Comment by phanikumv** [comment]
Did a quick POC and was able to launch aws-nuke on our AWS account, as shown below. Next step is to integrate it with our staging airflow instance and run it daily at specified period of the day.


``
(base) phanikv@Phanis-MacBook-Pro nuke % ./aws-nuke-v2.17.0-darwin-amd64 -c config.yml --profile default
aws-nuke version v2.17.0

Do you really want to nuke the account with the ID 475538383708 and the alias 'aws-oss-main'?
Do you want to continue? Enter account alias to continue.
> aws-oss-main

WARN[0116] skipping request: DNS lookup failed for globalaccelerator.us-east-2.amazonaws.com; assuming it does not exist in this region 
us-east-2 - KMSAlias - alias/aws/dynamodb - [Name: "alias/aws/dynamodb"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/ebs - [Name: "alias/aws/ebs"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/elasticfilesystem - [Name: "alias/aws/elasticfilesystem"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/es - [Name: "alias/aws/es"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/glue - [Name: "alias/aws/glue"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/kinesisvideo - [Name: "alias/aws/kinesisvideo"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/rds - [Name: "alias/aws/rds"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/redshift - [Name: "alias/aws/redshift"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/redshifttest - [Name: "alias/aws/redshifttest"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/s3 - [Name: "alias/aws/s3"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/ssm - [Name: "alias/aws/ssm"] - cannot delete AWS alias
us-east-2 - KMSAlias - alias/aws/xray - [Name: "alias/aws/xray"] - cannot delete AWS alias
WARN[0118] skipping request: DNS lookup failed for globalaccelerator.us-east-2.amazonaws.com; assuming it does not exist in this region 
us-east-2 - AthenaWorkGroup - primary - [ARN: "arn:aws:athena:us-east-2:475538383708:workgroup/primary", Name: "primary"] - cannot delete primary athena work group
us-east-2 - EC2DHCPOption - dopt-059bbfad9f92078f6 - [DefaultVPC: "true"] - would remove
us-east-2 - GlueDatabase - default - would remove
us-east-2 - GlueDatabase - providers_db - would remove
WARN[0119] skipping request: DNS lookup failed for worklink.us-east-2.amazonaws.com; assuming it does not exist in this region 
us-east-2 - Route53ResolverRule - rslvr-autodefined-rr-internet-resolver (Internet Resolver) - [ID: "rslvr-autodefined-rr-internet-resolver", Name: "Internet Resolver"] - Filtering DomainName "."
us-east-2 - CloudFormationStack - eksctl-providers-team-eks-cluster-cluster - [CreationTime: "2022-04-14T18:00:22Z", LastUpdatedTime: "2022-04-14T18:00:22Z", Name: "eksctl-providers-team-eks-cluster-cluster", tag:alpha.eksctl.io/cluster-name: "providers-team-eks-cluster", tag:alpha.eksctl.io/eksctl-version: "0.92.0", tag:eksctl.cluster.k8s.io/v1alpha1/cluster-name: "providers-team-eks-cluster"] - would remove
us-east-2 - CloudWatchEventsTarget - Rule: AutoScalingManagedRule Target ID: autoscaling - would remove
INFO[0121] TagOptionNotMigratedException: TagOption Migration not complete 
us-east-2 - EC2Subnet - subnet-0d9b95c165c481808 - [DefaultForAz: "true", DefaultVPC: "true"] - would remove
us-east-2 - EC2Subnet - subnet-0815bfc21b6b4c71a - [DefaultForAz: "true", DefaultVPC: "true"] - would remove
us-east-2 - EC2Subnet - subnet-083a7e2a7c3dc82a0 - [DefaultForAz: "true", DefaultVPC: "true"] - would remove
us-east-2 - CloudWatchEventsRule - Rule: AutoScalingManagedRule - would remove
us-east-2 - EMRCluster - j-LXCZH4ZP5GHO - already terminated
``

**Comment by phanikumv** [comment]
AWS nuke ran fine on staging, need to check why it is failing on astro-cloud

**Comment by phanikumv** [comment]
AWS nuke ran fine on astro-cloud #323

**Comment by pankajkoti** [comment]
In case of GCP, we are using following resources:
1. GCS Buckets
2. Compute Instances
3. DataProc Cluster
4. Kubernetes cluster

Upon exploring available options, there is no direct library available to destroy those resources as per our need. We will need to write a custom shell script to list and destroy the resources.

We also need to check how to securely store the service account key file json needed for authentication & to be used by the shell script as part of our DAG/CI.

**Comment by pankajkoti** [comment]
Phani unblocked me with assigning permissions to the service account for GCP nuke. 

Started writing DAG for nuking. Connection based approach does not work as secrets do not come. Using Variable based approach. Able to authenticate and execute gcloud commands with that

**Comment by pankajkoti** [comment]
Need additional permissions to create above resources to my dev account (this is different from the service account) to be able to test the nuke.

**Comment by pankajkoti** [comment]
PR is ready for the above, tested nuking the resources created in GCP example DAGs: https://github.com/astronomer/astronomer-providers/pull/563

**Comment by pankajkoti** [comment]
GCP nuke DAG is deployed and it is running fine on Astro Cloud.

**Comment by bharanidharan14** [comment]
Found Azure nuke in open source https://github.com/ekristen/azure-nuke Originally based on the source code from aws-nuke fork and aws-nuke original need suggestion on this

**Comment by bharanidharan14** [comment]
In Azure these are the following resource used as part of example DAG
1. Azure Data factory pipeline
2. Service Bus
3. Data Storage blob container

there is different ways to terminate the resource in azure
1. Use Azure cli commands in the custom shell script
2. Use Azure PowerShell commands 
2. Use the Open source library for Azure nuke https://github.com/ekristen/azure-nuke

**Comment by bharanidharan14** [comment]
Started with Azure nuke with https://github.com/ekristen/azure-nuke added example DAG, nuke-config.yml file, Testing is pending

**Comment by pankajkoti** [comment]
Nuke is complete for AWS & GCP. For Azure, we have a separate ticket: https://app.zenhub.com/workspaces/airflow-providers-team-61a89ee7fbf048001dc8378d/issues/astronomer/astronomer-providers/570
